### PR TITLE
Replace `Closable` with `Once`

### PR DESCRIPTION
**Pull request by kitcambridge**
I found this helpful while working on #179...there are a few more places where the `Closable` pattern is handy, but the explicit initialization step (`a.Closable.CloserOnce = a`) is cumbersome. It also requires an explicit constructor, whereas `Once` can be used without initialization. Forgetting to initialize a `CloserOnce` makes for a nice surprise when the shutdown logic tries to call `Close()` and panics.

This also adds tests for `Once`. It really is identical to `sync.Once`, save for accepting a function that returns an error, and exposing `IsDone()`.

### CODE REVIEW

File: src/github.com/mozilla-services/pushgo/simplepush/app.go [hunk 1]
```diff
@@ -294,7 +293,11 @@ func (a *Application) closeClients() {
 	}
 }
 
-func (a *Application) CloseOnce() error {
+func (a *Application) Close() error {
+	return a.closeOnce.Do(a.close)
+}
+
+func (a *Application) close() error {
```
Comment (jrconlin) [inline src/github.com/mozilla-services/pushgo/simplepush/app.go]: Kind of a nit, but we only ever call close() from Close(), and then we're doing some stack juggling to call the function. I like the pattern of Closer, but it seems like we could just call ``` if !a.closeOnce.IsClosed() { 
    ...  
    a.closeOnce.Close()
} ```
and do the same without the stack push/pops.

Comment (kitcambridge) [inline src/github.com/mozilla-services/pushgo/simplepush/app.go]: Agreed; there's a lot of indirection. `if !a.closeOnce.IsDone() { ... a.closeOnce.Close() }` doesn't look safe for concurrent calls, but I think a closure would work.

File: src/github.com/mozilla-services/pushgo/simplepush/structs.go [hunk 2]
```diff
@@ -29,48 +28,32 @@ type PushCommand struct {
...
+	return ws.uaid
 }
 
 func (ws *PushWS) SetUAID(uaid string) {
-	ws.uaidLock.Lock()
 	ws.uaid = uaid
-	ws.uaidLock.Unlock()
 }
 
 func (ws *PushWS) IsClosed() (closed bool) {
```
Comment (jrconlin) [inline src/github.com/mozilla-services/pushgo/simplepush/structs.go]: From what I can see in the code, we're only calling this from Serv.Bye(). If we moved the Unregister and cleanup after the sock.Close() we might be able to remove that check as well.

Comment (kitcambridge) [inline src/github.com/mozilla-services/pushgo/simplepush/structs.go]: The reason behind `IsClosed()` is convoluted...short version: it mitigates a race condition caused by reconnecting clients. Here's the long version; maybe it'll make it easier to come up with a proper solution:

The issue crops up when a client drops its connection ("connection A"), then reconnects with the same ID ("connection B"). A and B are handled by separate goroutines, each with an active [`Run()` loop](https://github.com/mozilla-services/pushgo/blob/dev/src/github.com/mozilla-services/pushgo/simplepush/handlers_socket.go#L138). When B connects, we assume A is stale and should be replaced. So we [close A](https://github.com/mozilla-services/pushgo/blob/dev/src/github.com/mozilla-services/pushgo/simplepush/worker.go#L479-485), which unblocks its `Run()` loop and [calls `Bye()`](https://github.com/mozilla-services/pushgo/blob/dev/src/github.com/mozilla-services/pushgo/simplepush/handlers_socket.go#L131) in a `defer` block. This cleanup also runs if the client gracefully closes its connection.

Because the cleanup happens in A's goroutine, B's goroutine doesn't know when `Bye()` has returned—it only closes A, then adds B to the client map. Since A and B share the same device ID, there's a race where B is added to the map before goroutine A cleans up. In that case, when goroutine A calls `Bye()`, it'll remove the newly-added B from the map.

The `IsClosed()` hack ensures we only remove a client from the map once. When goroutine B closes stale connection A, it removes A from the map and sets `IsClosed()` to true. When goroutine A then goes to clean up, it'll see the connection has already been closed and do nothing.

Does that make sense? Is there a better way out?

Comment (bbangert) [inline src/github.com/mozilla-services/pushgo/simplepush/structs.go]: Yes, the better way is for RemoveClient to take two args. The uaid *and* the client, and RemoveClient will only remove the uaid if the client objects match. This way if a new one is added, old RemoveClient calls will not remove the new uaid/client.

Comment (kitcambridge) [inline src/github.com/mozilla-services/pushgo/simplepush/structs.go]: That'll do it. :grin: Thanks, @bbangert!

Comment (kitcambridge) [inline src/github.com/mozilla-services/pushgo/simplepush/structs.go]: Going to make a separate PR for this; `Client` is currently an ephemeral type that's only used for the map, but we can clean up the duplication among `PushWS`, `WorkerWS`, `Socket`, and `Client` considerably now that we'll be holding onto `Client` objects.

File: src/github.com/mozilla-services/pushgo/simplepush/config.go [hunk 3]
```diff
@@ -396,21 +396,21 @@ func LoadApplication(configFile ConfigFile, env envconf.Environment,
 			return serv, nil
 		},
 		PluginSocket: func(app *Application) (HasConfigStruct, error) {
-			h := NewSocketHandler()
+			h := new(SocketHandler)
```
Comment (bbangert) [inline src/github.com/mozilla-services/pushgo/simplepush/config.go]: Why the switch to new()? The main reason so much code sticks with a function to return it is that way if the initialization requirement changes, you don't have to go find every new(Whatever), you can just change NewWhatever and all the existing code *just works*. Is there some hidden reason for this change?

**Comment by bbangert** [comment]
Looks good, will need to be re-based.

**Comment by bbangert** [comment]
r+
